### ISSUE TITLE
Error encoding a character on webhook call

### ISSUE THREAD

**Issue by tzaeru** [buildbot/buildbot#4457]
Hi,

I apologize beforehand for the lack of in-depth error trace, but I don't seem to find any. In short, I have a Bitbucket webhook for Buildbot set and when the webhook is triggered, this is the response returned to Bitbucket: `'ascii' codec can't encode character u'\xe4' in position 10: ordinal not in range(128)`

In master.cfg, my webhook looks like this:
```python
c['www']['change_hook_dialects'] = {'bitbucketcloud': True}
c['www']['change_hook_auth'] = [strcred.makeChecker("file:changehook.passwd")]
```

My scheduler looks like this (and it might be wrong - on some other commit I didn't get the above error but the build still didn't trigger?):

```python
c['schedulers'].append(schedulers.SingleBranchScheduler(
                name="Staging API",
                change_filter=util.ChangeFilter(project='<group>/<project>', branch='master'),
                #change_filter=util.ChangeFilter(project='<group>/<project>', filter_fn = lambda: True),
                treeStableTimer=3,
                builderNames=["buildstaging"]))
```
`twistd.log` seems to have no error message related to the encoding problem. The only thing in there is this:
```
2018-11-29 12:15:46+0200 [_GenericHTTPChannelProtocol,0,104.192.136.191] Payload: {[lots of data]}
2018-11-29 12:15:46+0200 [_GenericHTTPChannelProtocol,0,104.192.136.191] Processing event X-Event-Key: repo:push
```

I'm omitting the full payload since I'm not sure if there's anything that I shouldn't share in there. All strings appear to be unicode strings, though curiously unicode characters like my name are nevertheless mangled: `u'display_name': u'Jalmari Ik\xe4valko',`. Interestingly enough, what would be an `ä` there corresponds to the character index 10 of the string and a co-worker's, whose name contains only valid ascii characters, commits did not result in this error. So I think this error might happen because of my non-ascii name. Perhaps Buildbot attempts to convert the unicode strings to ascii strings and fails when it encounters that non-ascii character?

**Comment by p12tic** [comment]
Hey, could you share an anonymized payload? E.g. replace continuous spans of uninteresting ASCII characters with something like `z`.

**Comment by tzaeru** [comment]
`Payload: {u'push': {u'changes': [{u'forced': False, u'old': {u'name': u'master', u'links': {u'commits': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commits/master'}, u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/refs/branches/master'}, u'html': {u'href': u'https://bitbucket.org/x/repo/branch/master'}}, u'default_merge_strategy': u'merge_commit', u'merge_strategies': [u'merge_commit', u'squash', u'fast_forward'], u'type': u'branch', u'target': {u'hash': u'97db54369811574b73706936014e9307d2721971', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/97db54369811574b73706936014e9307d2721971'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/97db54369811574b73706936014e9307d2721971'}}, u'author': {u'raw': u'x <x.x@x.com>', u'type': u'author', u'user': {u'username': u'x', u'display_name': u'x', u'account_id': u'5bcd7918653a520f925a91ec', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/users/x'}, u'html': {u'href': u'https://bitbucket.org/x/'}, u'avatar': {u'href': u'https://bitbucket.org/account/x/avatar/'}}, u'nickname': u'x', u'type': u'user', u'uuid': u'{0f8d54c6-5303-47b9-beef-036b1e49864a}'}}, u'summary': {u'raw': u"Merge branch 'without-stream'\n", u'markup': u'markdown', u'html': u"<p>Merge branch 'without-stream'</p>", u'type': u'rendered'}, u'parents': [{u'hash': u'21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e'}}}, {u'hash': u'40008674c0cbcff13320e192ce51f24a2caf678d', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/40008674c0cbcff13320e192ce51f24a2caf678d'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/40008674c0cbcff13320e192ce51f24a2caf678d'}}}], u'date': u'2018-12-04T08:19:59+00:00', u'message': u"Merge branch 'without-stream'\n", u'type': u'commit'}}, u'links': {u'commits': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commits?include=d0b9fd82311d2957d283a13650aa9f669bb250d7&exclude=97db54369811574b73706936014e9307d2721971'}, u'html': {u'href': u'https://bitbucket.org/x/repo/branches/compare/d0b9fd82311d2957d283a13650aa9f669bb250d7..97db54369811574b73706936014e9307d2721971'}, u'diff': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/diff/d0b9fd82311d2957d283a13650aa9f669bb250d7..97db54369811574b73706936014e9307d2721971'}}, u'created': False, u'commits': [{u'hash': u'd0b9fd82311d2957d283a13650aa9f669bb250d7', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/d0b9fd82311d2957d283a13650aa9f669bb250d7'}, u'comments': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/d0b9fd82311d2957d283a13650aa9f669bb250d7/comments'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/d0b9fd82311d2957d283a13650aa9f669bb250d7'}, u'diff': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/diff/d0b9fd82311d2957d283a13650aa9f669bb250d7'}, u'approve': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/d0b9fd82311d2957d283a13650aa9f669bb250d7/approve'}, u'statuses': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/d0b9fd82311d2957d283a13650aa9f669bb250d7/statuses'}}, u'author': {u'raw': u'tzaeru <jalmari.ikavalko@x.com>', u'type': u'author', u'user': {u'username': u'tzaeru', u'display_name': u'Jalmari Ik\xe4valko', u'account_id': u'557058:c11a1f06-0cac-43f1-aa0c-eb1de598466d', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/users/tzaeru'}, u'html': {u'href': u'https://bitbucket.org/tzaeru/'}, u'avatar': {u'href': u'https://bitbucket.org/account/tzaeru/avatar/'}}, u'nickname': u'tzaeru', u'type': u'user', u'uuid': u'{263e2b2e-42c4-4fe5-a272-594dad596aae}'}}, u'summary': {u'raw': u"Merge branch 'master' of bitbucket.org:x/repo\n", u'markup': u'markdown', u'html': u"<p>Merge branch 'master' of bitbucket.org:x/repo</p>", u'type': u'rendered'}, u'parents': [{u'hash': u'b6d7c53052084d997dfdf5bb66ef55f091fdca53', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}}}, {u'hash': u'97db54369811574b73706936014e9307d2721971', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/97db54369811574b73706936014e9307d2721971'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/97db54369811574b73706936014e9307d2721971'}}}], u'date': u'2018-12-04T16:12:29+00:00', u'message': u"Merge branch 'master' of bitbucket.org:x/repo\n", u'type': u'commit'}, {u'hash': u'b6d7c53052084d997dfdf5bb66ef55f091fdca53', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'comments': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53/comments'}, u'patch': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/patch/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'diff': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/diff/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'approve': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53/approve'}, u'statuses': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53/statuses'}}, u'author': {u'raw': u'tzaeru <jalmari.ikavalko@x.com>', u'type': u'author', u'user': {u'username': u'tzaeru', u'display_name': u'Jalmari Ik\xe4valko', u'account_id': u'557058:c11a1f06-0cac-43f1-aa0c-eb1de598466d', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/users/tzaeru'}, u'html': {u'href': u'https://bitbucket.org/tzaeru/'}, u'avatar': {u'href': u'https://bitbucket.org/account/tzaeru/avatar/'}}, u'nickname': u'tzaeru', u'type': u'user', u'uuid': u'{263e2b2e-42c4-4fe5-a272-594dad596aae}'}}, u'summary': {u'raw': u'CI integration test\n', u'markup': u'markdown', u'html': u'<p>CI integration test</p>', u'type': u'rendered'}, u'parents': [{u'hash': u'21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/21377bb0ba8d2273b0fef0ec2b00a7df9f6e928e'}}}], u'date': u'2018-12-04T16:12:22+00:00', u'message': u'CI integration test\n', u'type': u'commit'}], u'truncated': False, u'closed': False, u'new': {u'name': u'master', u'links': {u'commits': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commits/master'}, u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/refs/branches/master'}, u'html': {u'href': u'https://bitbucket.org/x/repo/branch/master'}}, u'default_merge_strategy': u'merge_commit', u'merge_strategies': [u'merge_commit', u'squash', u'fast_forward'], u'type': u'branch', u'target': {u'hash': u'd0b9fd82311d2957d283a13650aa9f669bb250d7', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/d0b9fd82311d2957d283a13650aa9f669bb250d7'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/d0b9fd82311d2957d283a13650aa9f669bb250d7'}}, u'author': {u'raw': u'tzaeru <jalmari.ikavalko@x.com>', u'type': u'author', u'user': {u'username': u'tzaeru', u'display_name': u'Jalmari Ik\xe4valko', u'account_id': u'557058:c11a1f06-0cac-43f1-aa0c-eb1de598466d', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/users/tzaeru'}, u'html': {u'href': u'https://bitbucket.org/tzaeru/'}, u'avatar': {u'href': u'https://bitbucket.org/account/tzaeru/avatar/'}}, u'nickname': u'tzaeru', u'type': u'user', u'uuid': u'{263e2b2e-42c4-4fe5-a272-594dad596aae}'}}, u'summary': {u'raw': u"Merge branch 'master' of bitbucket.org:x/repo\n", u'markup': u'markdown', u'html': u"<p>Merge branch 'master' of bitbucket.org:x/repo</p>", u'type': u'rendered'}, u'parents': [{u'hash': u'b6d7c53052084d997dfdf5bb66ef55f091fdca53', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/b6d7c53052084d997dfdf5bb66ef55f091fdca53'}}}, {u'hash': u'97db54369811574b73706936014e9307d2721971', u'type': u'commit', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo/commit/97db54369811574b73706936014e9307d2721971'}, u'html': {u'href': u'https://bitbucket.org/x/repo/commits/97db54369811574b73706936014e9307d2721971'}}}], u'date': u'2018-12-04T16:12:29+00:00', u'message': u"Merge branch 'master' of bitbucket.org:x/repo\n", u'type': u'commit'}}}]}, u'actor': {u'username': u'tzaeru', u'display_name': u'Jalmari Ik\xe4valko', u'account_id': u'557058:c11a1f06-0cac-43f1-aa0c-eb1de598466d', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/users/tzaeru'}, u'html': {u'href': u'https://bitbucket.org/tzaeru/'}, u'avatar': {u'href': u'https://bitbucket.org/account/tzaeru/avatar/'}}, u'nickname': u'tzaeru', u'type': u'user', u'uuid': u'{263e2b2e-42c4-4fe5-a272-594dad596aae}'}, u'repository': {u'scm': u'git', u'website': u'', u'name': u'repo', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/repositories/x/repo'}, u'html': {u'href': u'https://bitbucket.org/x/repo'}, u'avatar': {u'href': u'https://bytebucket.org/ravatar/%7B74e884ba-ed21-4dd9-af5f-f2876e5a8e4c%7D?ts=default'}}, u'project': {u'key': u'server', u'type': u'project', u'uuid': u'{54932634-05eb-4aed-9817-3afb6c573d9b}', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/teams/x/projects/server'}, u'html': {u'href': u'https://bitbucket.org/account/user/x/projects/server'}, u'avatar': {u'href': u'https://bitbucket.org/account/user/x/projects/server/avatar/32'}}, u'name': u'Uplause Server 2.0'}, u'full_name': u'x/repo', u'owner': {u'username': u'x', u'display_name': u'Company Ltd', u'type': u'team', u'uuid': u'{764eb3fa-4091-4406-8482-28cb0a65a5c5}', u'links': {u'self': {u'href': u'https://api.bitbucket.org/2.0/teams/x'}, u'html': {u'href': u'https://bitbucket.org/x/'}, u'avatar': {u'href': u'https://bitbucket.org/account/x/avatar/'}}}, u'type': u'repository', u'is_private': True, u'uuid': u'{74e884ba-ed21-4dd9-af5f-f2876e5a8e4c}'}}`

**Comment by rodrigc** [comment]
@tzaeru would you be able to change your Buildbot master to run with Python 3, and see if the same problem occurs?  This looks like a unicode encoding issue.  Python 3 deals much better with these types of unicode encoding issues.

Tracing these things down on Python 2 is hard.

**Comment by tzaeru** [comment]
Well, now I feel a little bit dumb..

Changed to Python 3 and the error doesn't seem to happen anymore, not on the first try at least. Webhook with commit info with my unicode-including real world name returned 200, build triggered, and everything appears to work as intended. The silly part is that originally I had set up Python 3 sandbox for buildbot, but when I restarted it after config changes, I forgot to switch to that sandbox so I had accidentally started it with Python 2.
